gammapy/modeling/selection.py

```diff
@@ -60,10 +61,43 @@ def ts_threshold(self):
         This assumes that the TS follows a chi squared distribution
         with a number of degree of freedom equal to `n_free_parameters`.
         """
-        return sigma_to_ts(self.n_sigma, self.n_free_parameters)
+        return np.sign(self.n_sigma) * sigma_to_ts(self.n_sigma, self.n_free_parameters)
+
+    def ts_frozen(self, datasets):
+        """Perform the alternative hypothesis testing with all parameters frozen,
+        assuming that the non-null model is the true model.
+        If the assumption is true the frozen_ts should tend to the asimov_ts.
+        """
+        stat = datasets.stat_sum()
+        object_cache, prev_pars = self._apply_null_hypothesis(datasets)
+        stat_null = datasets.stat_sum()
+        self._restore_status(datasets, object_cache, prev_pars)
+        return stat_null - stat
+
+    def ts_asimov(self, datasets):
+        """Perform the alternative hypothesis testing with all parameters frozen,
+        for the asimov dataset i.e the non-null model is the true model.
+        """
+        counts_cahe = [d.counts for d in datasets]
+        for d in datasets:
+            d.counts = d.npred()
+
+        ts = self.ts_frozen(datasets)
+
+        for kd, d in enumerate(datasets):
+            d.counts = counts_cahe[kd]
+        return ts
+
+    def ts(self, datasets):
```

Comment:
simplified to self.run(datasets, apply_selection=False)['ts']

Reply:
Top... It is really necessary to redo the fit? Can you not cache the result?
I am not sure about the exact data flow of this estimator... ie if as usual we call the run function and then retrieve the result, or if you foresee a different usage

Reply:
Here .ts() is just a convenience function if you don't want to apply the selection immediately.
For the catalog I compute the TS detection of all the source first and I remove the ones below threshold all together only afterward (so for each source tested the global model is the uncut catalog and I can compare to TS_frozen to identify potential confusion effects). While for ts curvature, extension, shape... I apply the selection immediately, to avoid caching all the alternative models combination.